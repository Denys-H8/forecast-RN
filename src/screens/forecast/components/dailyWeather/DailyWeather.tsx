import React from 'react';
import { View, FlatList, TouchableOpacity } from 'react-native';

import constants from 'src/constants';
import CustomText from 'src/components/text';
import Icon from 'react-native-vector-icons/Fontisto';
import TemperatureScale from '../atomic/temperatureScale';
import { getCurrentDate, getWeatherIconName, timestampToDate } from 'src/helpers';
import { useSelector } from 'react-redux';
import { dayPeriod } from 'src/store/settings/selectors';
import { styles } from './dailyWeather.styles';
import { rootNavigation } from 'src/services';
import { weatherStatsSelector } from 'src/store/weather/selectors';
import { TDaily } from 'src/store/weather/reducer';

export const handleDayPress = (dateIn: string) => {
  const currentDate = getCurrentDate()[0];
  const dateOut = dateIn === currentDate ? 'today' : dateIn;

  rootNavigation.navigate('Weather', { date: dateOut });
};

export const getDay = (dateLine: string) => {
  const date = new Date(dateLine);
  const day = date.getDay();
  const currentDate = Date.now();
  const currentDay = new Date(currentDate).getDay();

  const days = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

  return days[currentDay] === days[day] ? 'Today' : days[day];
};

const DailyWeather: React.FC = () => {
  const dayPeriodValue = useSelector(dayPeriod);
  const weatherStats = useSelector(weatherStatsSelector);

  const weather = weatherStats.daily.slice(0, dayPeriodValue);
  const weatherResult: { [key: string]: string[]; weatherDescription: string[] }[] = [];

  weather.forEach((el: TDaily) => {
    weatherResult.push({
      [timestampToDate(el.dt)]: [el.temp.min.toFixed(0), el.temp.max.toFixed(0)],
      weatherDescription: [el.weather[0].description],
    });
  });

  const renderItem = (item: any) => {
    return (
      <TouchableOpacity
        onPress={() => handleDayPress(Object.keys(item)[0])}
        style={styles.itemContainer}
        testID={`DailyWeather.TouchableOpacity.Day.${Object.keys(item)[0]}`}
      >
        <CustomText style={styles.day}>{getDay(Object.keys(item)[0])}</CustomText>
        <View style={styles.iconScaleContainer}>
          <Icon
            name={getWeatherIconName(item.weatherDescription[0])}
            size={35}
            color={constants.SUB_COLOR}
            style={styles.icon}
          />
          <TemperatureScale
            minTemperature={Math.min(...item[Object.keys(item)[0]])}
            maxTemperature={Math.max(...item[Object.keys(item)[0]])}
          />
        </View>
      </TouchableOpacity>
    );
  };

  const keyExtractor = (item: {}) => Object.keys(item)[0];

  return (
    <View style={styles.container}>
      <FlatList
        keyExtractor={keyExtractor}
        data={weatherResult}
        ItemSeparatorComponent={() => <View style={styles.separator}></View>}
        renderItem={({ item }) => renderItem(item)}
      />
    </View>
  );
};

export default DailyWeather;
