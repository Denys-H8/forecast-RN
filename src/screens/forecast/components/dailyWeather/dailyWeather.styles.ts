import { StyleSheet } from 'react-native';

import constants from 'src/constants';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  itemContainer: {
    flexDirection: 'row',
    paddingLeft: 33,
    paddingVertical: 22,
    alignItems: 'center',
    justifyContent: 'space-between',
    borderBottomWidth: 2,
    borderBottomColor: constants.SUB_COLOR,
  },
  day: {
    color: constants.MAIN_COLOR,
    fontSize: 28,
    lineHeight: 32,
    width: 80,
  },
  degreesContainer: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    width: 80,
  },
  number: {
    fontSize: 36,
    lineHeight: 42,
    color: constants.MAIN_COLOR,
  },
  celsius: {
    fontSize: 18,
    lineHeight: 18,
    color: constants.MAIN_COLOR,
  },
  iconScaleContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'flex-start',
  },
  icon: {
    width: 50,
    textAlign: 'center',
    marginLeft: 15,
    marginRight: 20,
  },
  separator: {
    height: 2,
    backgroundColor: constants.SUB_COLOR,
  },
});
