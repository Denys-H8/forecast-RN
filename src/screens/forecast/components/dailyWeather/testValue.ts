export const weatherStatsValue = {
  lat: 50,
  lon: 36.25,

  current: {
    dt: 1619098295,
    sunrise: 1619058476,
    sunset: 1619109522,
    temp: 17.89,
    feels_like: 16.57,
    pressure: 1018,
    humidity: 32,
    clouds: 40,
    visibility: 10000,
    wind_speed: 6,
    wind_deg: 200,
    weather: [
      {
        id: 802,
        main: 'Clouds',
        description: 'scattered clouds',
        icon: '03d',
      },
    ],
  },
  hourly: [
    {
      dt: 1619096400,
      temp: 17.31,
      feels_like: 16.01,
      pressure: 1018,
      humidity: 35,
      dew_point: 1.72,
      uvi: 0.77,
      clouds: 52,
      visibility: 10000,
      wind_speed: 2.46,
      wind_deg: 232,
      wind_gust: 4.31,
      weather: [
        {
          id: 803,
          main: 'Clouds',
          description: 'broken clouds',
          icon: '04d',
        },
      ],
      pop: 0,
    },
    {
      dt: 1619100000,
      temp: 17.89,
      feels_like: 16.57,
      pressure: 1018,
      humidity: 32,
      dew_point: 0.98,
      uvi: 0.4,
      clouds: 40,
      visibility: 10000,
      wind_speed: 2.79,
      wind_deg: 217,
      wind_gust: 4.62,
      weather: [
        {
          id: 802,
          main: 'Clouds',
          description: 'scattered clouds',
          icon: '03d',
        },
      ],
      pop: 0,
    },
  ],
  daily: [
    {
      dt: 1619082000,
      sunrise: 1619058476,
      sunset: 1619109522,
      moonrise: 1619086740,
      moonset: 1619052900,
      moon_phase: 0.32,
      temp: {
        day: 15.18,
        min: 6.02,
        max: 17.89,
        night: 12.24,
        eve: 17.24,
        morn: 6.02,
      },
      feels_like: {
        day: 13.9,
        night: 4.6,
        eve: 15.96,
        morn: 4.6,
      },
      pressure: 1020,
      humidity: 44,
      dew_point: 3.24,
      wind_speed: 3.61,
      wind_deg: 218,
      wind_gust: 5.73,
      weather: [
        {
          id: 500,
          main: 'Rain',
          description: 'light rain',
          icon: '10d',
        },
      ],
      clouds: 16,
      pop: 0.32,
      rain: 0.26,
      uvi: 4.19,
    },
    {
      dt: 1615084000,
      sunrise: 1619058476,
      sunset: 1619109522,
      moonrise: 1619086740,
      moonset: 1619052900,
      moon_phase: 0.32,
      temp: {
        day: 15.18,
        min: 6.02,
        max: 17.89,
        night: 12.24,
        eve: 17.24,
        morn: 6.02,
      },
      feels_like: {
        day: 13.9,
        night: 4.6,
        eve: 15.96,
        morn: 4.6,
      },
      pressure: 1020,
      humidity: 44,
      dew_point: 3.24,
      wind_speed: 3.61,
      wind_deg: 218,
      wind_gust: 5.73,
      weather: [
        {
          id: 500,
          main: 'Rain',
          description: 'light rain',
          icon: '10d',
        },
      ],
      clouds: 16,
      pop: 0.32,
      rain: 0.26,
      uvi: 4.19,
    },
  ],
};
