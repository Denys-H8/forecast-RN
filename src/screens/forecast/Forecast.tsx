import React from 'react';
import { View } from 'react-native';

import DailyWeather from './components/dailyWeather';
import { styles } from './forecast.styles';

const Forecast: React.FC = () => {
  return (
    <View style={styles.container}>
      <DailyWeather />
    </View>
  );
};

export default Forecast;
