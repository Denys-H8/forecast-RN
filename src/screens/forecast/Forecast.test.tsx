import React from 'react';
import { render } from '@testing-library/react-native';
import Forecast from './Forecast';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { weatherStatsValue } from './components/dailyWeather/testValue';

const mockStore = configureStore([]);

const store = mockStore({
  settingsReducer: {
    dayPeriod: 4,
  },
  weatherReducer: {
    weatherStats: weatherStatsValue,
  },
});

describe('Forecast ', () => {
  Date.now = jest.fn(() => 1619098295 * 1000);

  test('should render correctly', () => {
    const { toJSON } = render(
      <Provider store={store}>
        <Forecast />
      </Provider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
