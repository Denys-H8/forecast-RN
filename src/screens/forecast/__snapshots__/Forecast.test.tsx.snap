// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast  should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#191C39",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      data={
        Array [
          Object {
            "2021-04-22": Array [
              "6",
              "18",
            ],
            "weatherDescription": Array [
              "light rain",
            ],
          },
          Object {
            "2021-03-07": Array [
              "6",
              "18",
            ],
            "weatherDescription": Array [
              "light rain",
            ],
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "borderBottomColor": "#8B8EA5",
                "borderBottomWidth": 2,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "opacity": 1,
                "paddingLeft": 33,
                "paddingVertical": 22,
              }
            }
            testID="DailyWeather.TouchableOpacity.Day.2021-04-22"
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "WorkSans-Regular",
                  },
                  Object {
                    "color": "#fff",
                    "fontSize": 28,
                    "lineHeight": 32,
                    "width": 80,
                  },
                ]
              }
            >
              Today
            </Text>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#8B8EA5",
                      "fontSize": 35,
                    },
                    Object {
                      "marginLeft": 15,
                      "marginRight": 20,
                      "textAlign": "center",
                      "width": 50,
                    },
                    Object {
                      "fontFamily": "Fontisto",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "WorkSans-Regular",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 32,
                          "lineHeight": 36,
                        },
                      ]
                    }
                  >
                    6
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "WorkSans-Regular",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 18,
                          "lineHeight": 18,
                        },
                      ]
                    }
                  >
                    0
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#8B8EA5",
                        "borderRadius": 15,
                        "height": 12,
                        "marginHorizontal": 5,
                        "maxWidth": 85,
                      },
                      Object {
                        "width": 36,
                      },
                    ]
                  }
                />
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "WorkSans-Regular",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 32,
                          "lineHeight": 36,
                        },
                      ]
                    }
                  >
                    18
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "WorkSans-Regular",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 18,
                          "lineHeight": 18,
                        },
                      ]
                    }
                  >
                    0
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Object {
                "backgroundColor": "#8B8EA5",
                "height": 2,
              }
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "borderBottomColor": "#8B8EA5",
                "borderBottomWidth": 2,
                "flexDirection": "row",
                "justifyContent": "space-between",
                "opacity": 1,
                "paddingLeft": 33,
                "paddingVertical": 22,
              }
            }
            testID="DailyWeather.TouchableOpacity.Day.2021-03-07"
          >
            <Text
              style={
                Array [
                  Object {
                    "fontFamily": "WorkSans-Regular",
                  },
                  Object {
                    "color": "#fff",
                    "fontSize": 28,
                    "lineHeight": 32,
                    "width": 80,
                  },
                ]
              }
            >
              Sun
            </Text>
            <View
              style={
                Object {
                  "flex": 1,
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#8B8EA5",
                      "fontSize": 35,
                    },
                    Object {
                      "marginLeft": 15,
                      "marginRight": 20,
                      "textAlign": "center",
                      "width": 50,
                    },
                    Object {
                      "fontFamily": "Fontisto",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "WorkSans-Regular",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 32,
                          "lineHeight": 36,
                        },
                      ]
                    }
                  >
                    6
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "WorkSans-Regular",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 18,
                          "lineHeight": 18,
                        },
                      ]
                    }
                  >
                    0
                  </Text>
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#8B8EA5",
                        "borderRadius": 15,
                        "height": 12,
                        "marginHorizontal": 5,
                        "maxWidth": 85,
                      },
                      Object {
                        "width": 36,
                      },
                    ]
                  }
                />
                <View
                  style={
                    Object {
                      "alignItems": "flex-start",
                      "flexDirection": "row",
                    }
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "WorkSans-Regular",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 32,
                          "lineHeight": 36,
                        },
                      ]
                    }
                  >
                    18
                  </Text>
                  <Text
                    style={
                      Array [
                        Object {
                          "fontFamily": "WorkSans-Regular",
                        },
                        Object {
                          "color": "#fff",
                          "fontSize": 18,
                          "lineHeight": 18,
                        },
                      ]
                    }
                  >
                    0
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
